cpp/src/sort/rank.cu

```diff
@@ -0,0 +1,220 @@
+      rmm::device_vector<double> max_sorted(size, 0);
+      // algorithm: reduce_by_key(dense_rank, 1, n, max)
+      thrust::reduce_by_key(
+          rmm::exec_policy(stream)->on(stream), dense_rank_sorted.begin(),
+          dense_rank_sorted.end(), thrust::make_counting_iterator<double>(1),
+          thrust::make_discard_iterator(), max_sorted.begin(),
+          thrust::equal_to<double>{}, thrust::maximum<double>{});
+      auto sorted_max_rank = thrust::make_transform_iterator(
+          dense_rank_sorted.begin(), [max_rank = max_sorted.begin()] 
+          __device__( auto i) -> double { return max_rank[i - 1]; });
+      thrust::scatter(rmm::exec_policy(stream)->on(stream), sorted_max_rank,
+                      sorted_max_rank + size,
+                      sorted_order_view.begin<size_type>(), rank_data);
+    } break;
+    case rank_method::AVERAGE: {
+      using MinCount = thrust::tuple<size_type, size_type>;
+      // k, k+1, .. k+n-1
+      // average = (n*k+ n*(n-1)/2)/n
+      // average = k + (n-1)/2 = min + (count-1)/2
+      // Calculate Min of ranks and Count of equal values
+      rmm::device_vector<MinCount> min_count(size);
+      thrust::reduce_by_key(
+          rmm::exec_policy(stream)->on(stream), dense_rank_sorted.begin(),
+          dense_rank_sorted.end(),
+          thrust::make_zip_iterator(
+              thrust::make_tuple(thrust::make_counting_iterator<size_type>(1),
+                                 thrust::make_constant_iterator<size_type>(1))),
+          thrust::make_discard_iterator(), min_count.begin(),
+          thrust::equal_to<size_type>{}, [] __device__(auto i, auto j) {
+            return MinCount{std::min(thrust::get<0>(i), thrust::get<0>(j)),
+                            thrust::get<1>(i) + thrust::get<1>(j)};
+          });
+      // Transform (MinRank, Count) to Average Rank.
+      auto avgit = thrust::make_transform_iterator(
+          min_count.begin(),
+          [] __device__(auto i) { // min+(count-1)/2
+            return static_cast<double>(thrust::get<0>(i)) +
+                   (static_cast<double>(thrust::get<1>(i)) - 1) / 2.0;
+          });
+      auto sorted_mean_rank = thrust::make_transform_iterator(
+          dense_rank_sorted.begin(),
+          [avgit] __device__(auto i) { return avgit[i - 1]; });
+      thrust::scatter(rmm::exec_policy(stream)->on(stream), sorted_mean_rank,
+                      sorted_mean_rank + size,
+                      sorted_order_view.begin<size_type>(), rank_data);
+    } break;
+    default:
+      CUDF_FAIL("Unexpected rank_method for rank()");
+    }
+    
+    // pct inplace transform
+    if (percentage) {
+      size_type const count = (_include_nulls == include_nulls::NO)
+                                          ? size - input_col.null_count()
+                                          : size;
+      auto drs = dense_rank_sorted.data().get();
+      bool is_dense = (method == rank_method::DENSE);
```

Comment:
Instead of capturing `is_dense` in the lambda, why not use two different lambdas depending on the value of `is_dense`?

Reply:
Why is this preferred?
it increases lines of code, creates 2 kernels instead of 1 kernel. 
`is_dense` will be a constant passed to kernel, it's fast and no divergence too.

Reply:
I guess it's just the DevTech in me that doesn't like the extra instructions that result from evaluating the `if`.

Reply:
I can empathise. I've faced situations where I can either make a bool a template parameter or do an if check in the kernel. The pros of latter is smaller compile time and less code. The former benefitting from lower operations in the kernel. In my experience, passing a single, never changing bool doesn't affect performance too much when the function is a temporary lambda like this one. But it should be avoided in a generic device function because of unforeseen slowdowns in other use cases.